internals/daemon/api_notices.go

```diff
@@ -0,0 +1,154 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"regexp"
+	"time"
+
+	"github.com/canonical/pebble/internals/overlord/state"
+)
+
+// A very loose regex to ensure client keys are in the form "domain.com/key"
+var clientKeyRegexp = regexp.MustCompile(`([a-z0-9-_]+\.)+[a-z0-9-_]+/.+`)
+
+func v1GetNotices(c *Command, r *http.Request, _ *UserState) Response {
+	query := r.URL.Query()
+
+	typeStr := query.Get("type")
+	noticeType := state.NoticeTypeFromString(typeStr)
+	if typeStr != "" && noticeType == "" {
```

Comment:
For better or worse, this is pretty typical for Go APIs: if the zero value is not meaningful, it's common to use is it as "not found" value. For example, [http.StatusText](https://pkg.go.dev/net/http#StatusText) and [unix.SignalNum](https://pkg.go.dev/golang.org/x/sys/unix#SignalNum).

Reply:
well, I don't think we should follow bad practice just because everyone else is doing it. :)

It would also make this check a lot cleaner and more readable.

Reply:
I don't consider it bad practice, though. Functions that return a single value are somewhat easier to deal with than multi-valued ones. (I do understand the disadvantage: they're less safe, because with `err`-returning ones you have to explicitly check the error.)

I also don't see how it would be a "lot cleaner and more readable", or why the current code is not clean or readable? What do you find confusing about it?

I grepped through the Go stdlib and found two functions in this class that. `http.StatusText` that I've already mentioned, and [`mime.TypeByExtension`](https://pkg.go.dev/mime#TypeByExtension), which also returns `""` in the not-found case. I think following stdlib precedent is a pretty compelling argument.

I didn't find any similar lookup functions that had an error return in the stdlib, though I may have missed them. All the functions with error returns seem to perform some OS or I/O operation. But those that are just simple lookups like this seem to not return error.

Reply:
Ok, well my concern is that with
```
noticeType := state.NoticeTypeFromString(typeStr)
if typeStr != "" && noticeType == "" {
```
it takes a while to understand what's happening here, it's not immediately clear, and you have to understand the internal details of the NoticeTypeFromString function to understand what's happening here (namely, that it returns "" when the input is invalid).

In contrast, with
```
noticeType, err := state.NoticeTypeFromString(typeStr)
if err != nil {
```
it's immediately clear that there was an error converting the string to a notice type.

Ultimately I suppose this is a matter of personal taste though.